nemo/collections/asr/parts/preprocessing/features.py

```diff
@@ -428,5 +476,175 @@ def forward(self, x, seq_len):
+        return x
+
+    def _daligraph_splice_frames(self, x, nfeatures, x_len, stacking=1, subsampling=1):
+        if stacking > 1:
+            seq = [x]
+            for n in range(1, stacking):
+                f = dali.fn.slice(x, start=n, shape=x_len, axes=(1,), out_of_bounds_policy='pad', fill_values=0)
+                seq.append(f)
+            x = dali.fn.cat(*seq, axis=0)
+            nfeatures = nfeatures * stacking
+        if subsampling > 1:
+            out_len = (x_len + subsampling - 1) // subsampling
+            m = dali.fn.transforms.scale(scale=[subsampling, 1], center=[0.5, 0])
+            x = dali.fn.reshape(x, rel_shape=[1, 1, -1], layout="HWC")  # Layout required by WarpAffine
+            x = dali.fn.warp_affine(
+                x, matrix=m, size=dali.fn.cat(nfeatures, out_len), interp_type=dali.types.INTERP_NN
+            )
+            x = dali.fn.reshape(x, rel_shape=[1, 1], layout="ft")
+        return x
+
+    def init_dali_pipeline(self, batch_size, device, device_id, num_threads=4):
+        pipe = dali.pipeline.Pipeline(
+            batch_size=batch_size,
+            num_threads=num_threads,
+            device_id=device_id,
+            prefetch_queue_depth=1,
+            exec_async=True,
+            exec_pipelined=True,
+        )
+        with pipe:
+            audio = dali.fn.external_source(name="input_signal", device=device)
+            audio_len = dali.fn.external_source(name="length", device='cpu')
+
+            if self.stft_pad_amount is not None:
+                audio = self._daligraph_reflect_pad(audio, audio_len, self.stft_pad_amount)
+                audio_len = audio_len + 2 * self.stft_pad_amount
+
+            spec_len = dali.fn.cast((audio_len // self.hop_length) + 1, dtype=dali.types.INT64)
+
+            # Additive gaussian noise (dither)
+            if self.training and self.dither > 0.0:
+                gaussian_noise = dali.fn.random.normal(device=device)
+                audio = audio + self.dither * gaussian_noise
+
+            # Preemphasis filter
+            if self.preemph > 0.0:
+                audio = dali.fn.preemphasis_filter(audio, preemph_coeff=self.preemph)
+
+            # Power spectrogram
+            spec = dali.fn.spectrogram(
+                audio,
+                nfft=self.n_fft,
+                power=self.mag_power,
+                window_fn=self.window_tensor_lst,
+                window_length=self.win_length,
+                window_step=self.hop_length,
+                center_windows=True,
+                reflect_padding=True,
+            )
+
+            # Spectrogram to Mel Spectrogram
+            spec = dali.fn.mel_filter_bank(
+                spec, sample_rate=self.sample_rate, nfilter=self.nfilt, freq_low=self.lowfreq, freq_high=self.highfreq,
+            )
+
+            # log features if required
+            if self.log:
+                eps = self.log_zero_guard_value_fn(torch.float32)
```

Comment:
This works under AMP too right?

Reply:
this pipeline works with float32. I think it should be no problem with AMP. I am not that familiar with it, so perhaps you can answer: Would the input still be float32 in case of AMP?

Reply:
Pytorch lightning would move the input to this class to cuda, then amp would activate if `trainer.precision=16`. Can we add a parallel jenkins test that does the same as the main one but with `trainer.precision=16` set? If the config doesnt have it, it might need `+trainer.precision=16`.